File: cpp-droid/engine.cpp
#include <stdio.h>
#include <string.h>
#include "engine.h"
#include "ast.h"
#include "lexer.cpp"
#include "parser.cpp"

void execute_sql(std::string command) {
    bool is_tokenize = false;
    bool is_ast = false;
    std::string sql = command;

    if (command.rfind("tokenize ", 9) == 0) {
        is_tokenize = true;
        sql = command.substr(9);
    } else if (command.rfind("ast ", 4) == 0) {
        is_ast = true;
        sql = command.substr(4);
    }

    TokenList list = tokenize(sql);
    AST_Node root = parse_statement(list);

    if (is_tokenize) {
        if (list.has_error) {
            std::cout << "Syntax error in command\n";
        } else {
            for (const auto& t : list.tokens) {
                std::cout << "[" << token_type_to_string(t.type) << " - " << t.token << "]\n";
            }
        }
        free_tokens(&list);
    } else if (is_ast) {
        if (root.has_error) {
            std::cout << "Invalid AST in command\n";
        }
        std::cout << "[ERROR:00100] AST Parser not implemented yet.\n";
    } else {
        std::cout << "[ERROR:00100] AST Parser not implemented yet.\n";
    }
}

File: rust-droid/engine.rs
use crate::lexer::tokenize;
use crate::parser::parse_statement;
use crate::ast::AstNode;
use crate::lexer::TokenList;


pub fn execute_sql(command: &str) {
    let mut is_tokenize = false;
    let mut is_ast = false;
    let mut sql_command = command;

    if sql_command.starts_with("tokenize ") {
        is_tokenize = true;
        sql_command = sql_command.strip_prefix("tokenize ").unwrap();
    } else if sql_command.starts_with("ast ") {
        is_ast = true;
        sql_command = sql_command.strip_prefix("ast ").unwrap();
    } 

    let tokens: TokenList = tokenize(sql_command);
    let ast: AstNode = parse_statement(&tokens);

    if is_tokenize {
        if tokens.has_error {
            println!("Syntax error in command");
        } else {
            for t in tokens.tokens {
                println!("[{:?} - {}]", t.tk_type, t.token);
            }
        }
    } else if is_ast {
        if ast.has_error {
            println!("Invalid AST in command");
        }   
        println!("[ERROR:00100] AST Parser not implemented yet.");
    } else {
        println!("[ERROR:00100] AST Parser not implemented yet.");
    }
}
